runtime/drivers/druid/druidsqldriver/druid_api_sql_driver.go

```diff
@@ -40,193 +55,285 @@ type sqlConnection struct {
+	default:
+		return true, fmt.Errorf("unexpected response: %v", obj)
+	}
+}
+
+func (c *sqlConnection) QueryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	// total sum is 126 seconds (sum(2*2^x) from 0 to 5 inclusive)
+	re := retrier.NewRetrier(6, 2*time.Second, &coordinatorHTTPCheck{
+		c: c,
+	})
+	return re.RunCtx(ctx, func(ctx2 context.Context) (driver.Rows, retrier.Action, error) {
+		dr := newDruidRequest(query, args)
+		b, err := json.Marshal(dr)
 		if err != nil {
+			return nil, retrier.Fail, err
+		}
+
+		bodyReader := bytes.NewReader(b)
+
+		context.AfterFunc(ctx, func() {
+			tctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			r, err := http.NewRequestWithContext(tctx, http.MethodDelete, c.dsn+"/"+dr.Context.SQLQueryID, http.NoBody)
+			if err != nil {
+				return
+			}
+
+			resp, err := c.client.Do(r)
+			if err != nil {
+				return
+			}
 			resp.Body.Close()
-			return nil, err
+		})
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.dsn, bodyReader)
+		if err != nil {
+			return nil, retrier.Fail, err
 		}
 
-		types := toStringArray(obj.([]any))
-
-		transformers := make([]func(any) (any, error), len(columns))
-		for i, c := range types {
-			transformers[i] = identityTransformer
-			switch c {
-			case "TINYINT":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return int8(v), nil
-					default:
-						return v, nil
-					}
-				}
-			case "SMALLINT":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return int16(v), nil
-					default:
-						return v, nil
-					}
-				}
-			case "INTEGER":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return int32(v), nil
-					default:
-						return v, nil
-					}
-				}
-			case "BIGINT":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return int64(v), nil
-					default:
-						return v, nil
-					}
-				}
-			case "FLOAT":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return float32(v), nil
-					case string:
-						return strconv.ParseFloat(v, 32)
-					default:
-						return v, nil
-					}
-				}
-			case "DOUBLE":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case string:
-						return strconv.ParseFloat(v, 64)
-					default:
-						return v, nil
-					}
-				}
-			case "REAL":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case string:
-						return strconv.ParseFloat(v, 64)
-					default:
-						return v, nil
-					}
+		req.Header.Add("Content-Type", "application/json")
+		resp, err := c.client.Do(req)
+		if err != nil {
+			// nolint:errorlint // there's no wrapping
+			if v, ok := err.(*url.Error); ok {
```

Comment:
We don't know how the library will change in the future. Better to use `errors.As` which is created for this.

Reply:
I mean `error.As` returns a bool but then after that I have to get the error message.
I agree that the library implementation can change but it will be incompatible changes for all clients (which is rare and a bad practice by the library). But anyway, if they change they can wrap it double time and this code fails again. So I should Unwrap recursively - which looks weird and clumsy for error handling.
So for example what can happen if the error is wrapped double time:

```go
type Er1 struct {
}

func (*Er1) Error() string {
	return "er1"
}

type wrapError struct {
	msg string
	err error
}

func (w *wrapError) Error() string {
	return w.msg
}

func (w *wrapError) Unwrap() error {
	return w.err
}

func ret() error {
	var e error
	e = &Er1{}
	e = &wrapError{
		msg: "extra info",
		err: e,
	}
	e = &wrapError{
		msg: "superextra info",
		err: e,
	}
	return e
}

var _ error = &Er1{}

func TestError(t *testing.T) {
	e := ret()
	e0 := &Er1{}
	require.True(t, errors.As(e, &e0))
	require.Equal(t, "er1", errors.Unwrap(e).Error())
}
```
```
        	            	expected: "er1"
        	            	actual  : "extra info"
```

Reply:
`errors.As` doesn't just return a bool, it also assigns the error – it handles the recursive unwrapping automatically.

Reply:
Changed.